execution/gethexec/express_lane_service.go

```diff
@@ -0,0 +1,323 @@
+	// Calculate the time to the next round
+	timeToNextRound := nextRoundStart.Sub(arrivalTime)
+	// Check if the arrival timestamp is within AUCTION_CLOSING_DURATION of TIME_TO_NEXT_ROUND
+	return timeToNextRound <= es.auctionClosing
+}
+
+// Sequence express lane submission skips validation of the express lane message itself,
+// as the core validator logic is handled in `validateExpressLaneTx“
+func (es *expressLaneService) sequenceExpressLaneSubmission(
+	ctx context.Context,
+	msg *timeboost.ExpressLaneSubmission,
+	publishTxFn func(
+		parentCtx context.Context,
+		tx *types.Transaction,
+		options *arbitrum_types.ConditionalOptions,
+		delay bool,
+	) error,
+) error {
+	es.Lock()
+	defer es.Unlock()
+	control, ok := es.roundControl.Get(msg.Round)
+	if !ok {
+		return timeboost.ErrNoOnchainController
+	}
+	// Check if the submission nonce is too low.
+	if msg.Sequence < control.sequence {
+		return timeboost.ErrSequenceNumberTooLow
+	}
+	// Check if a duplicate submission exists already, and reject if so.
+	if _, exists := es.messagesBySequenceNumber[msg.Sequence]; exists {
+		return timeboost.ErrDuplicateSequenceNumber
+	}
+	// Log an informational warning if the message's sequence number is in the future.
+	if msg.Sequence > control.sequence {
+		log.Warn("Received express lane submission with future sequence number", "sequence", msg.Sequence)
+	}
+	// Put into the sequence number map.
+	es.messagesBySequenceNumber[msg.Sequence] = msg
+
+	for {
+		// Get the next message in the sequence.
+		nextMsg, exists := es.messagesBySequenceNumber[control.sequence]
+		if !exists {
+			break
+		}
+		if err := publishTxFn(
+			ctx,
+			nextMsg.Transaction,
+			msg.Options,
+			false, /* no delay, as it should go through express lane */
+		); err != nil {
+			// If the tx failed, clear it from the sequence map.
+			delete(es.messagesBySequenceNumber, msg.Sequence)
+			return err
+		}
+		// Increase the global round sequence number.
+		control.sequence += 1
+	}
+	es.roundControl.Add(msg.Round, control)
+	return nil
+}
+
+func (es *expressLaneService) validateExpressLaneTx(msg *timeboost.ExpressLaneSubmission) error {
+	if msg == nil || msg.Transaction == nil || msg.Signature == nil {
+		return timeboost.ErrMalformedData
+	}
+	if msg.ChainId.Cmp(es.chainConfig.ChainID) != 0 {
+		return errors.Wrapf(timeboost.ErrWrongChainId, "express lane tx chain ID %d does not match current chain ID %d", msg.ChainId, es.chainConfig.ChainID)
+	}
+	if msg.AuctionContractAddress != es.auctionContractAddr {
+		return errors.Wrapf(timeboost.ErrWrongAuctionContract, "msg auction contract address %s does not match sequencer auction contract address %s", msg.AuctionContractAddress, es.auctionContractAddr)
+	}
+	if !es.currentRoundHasController() {
+		return timeboost.ErrNoOnchainController
+	}
+	currentRound := timeboost.CurrentRound(es.initialTimestamp, es.roundDuration)
+	if msg.Round != currentRound {
```

Comment:
I think we would want some buffering for next-round messages, to make sure winner can use their entire slot.

Reply:
I thought of a straight forward way to address this: https://github.com/OffchainLabs/nitro/pull/2788